### PR TITLE
feat: Add timestamp arithmetic functionality

### PR DESCRIPTION
**Pull request by jzaralim**
### Description 
This PR adds the functions `timestampadd` and `timestampsub`. Example usages are:

```
TIMESTAMPADD(MINUTES, 40, time)
TIMESTAMPSUB(HOURS, 10, time)
```

~~In order to do this, a new "type" was introduced, `INTERVAL`, that represents an interval of time. Users define them using an IntervalExpression, which consists of numeric expression and a WindowUnit. `INTERVAL` can only be used within a function and cannot be used as an actual data type (similar to intervals in MySQL).~~

### Testing done 
Unit/QTT tests for each of the functions

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs/developer-guide/ksqldb-reference/scalar-functions.md [hunk 1]
```diff
@@ -1151,6 +1151,28 @@ FROM_UNIXTIME(milliseconds)
...
 
+### TIMESTAMPADD
+
+Since: 0.16
+
+```sql
+TIMESTAMPADD(COL0, 5 MINUTES)
+```
+
+Adds an interval to a timestamp. Intervals are defined by an integral value and a supported
```
Comment (JimGalasyn) [inline docs/developer-guide/ksqldb-reference/scalar-functions.md:1162]:
```suggestion
Adds an interval to a timestamp. Intervals are defined by an integer value and a supported
```

File: docs/developer-guide/ksqldb-reference/scalar-functions.md [hunk 2]
```diff
@@ -1151,6 +1151,28 @@ FROM_UNIXTIME(milliseconds)
...
+
+### TIMESTAMPSUB
+
+Since: 0.16
+
+```sql
+TIMESTAMPSUB(COL0, 5 MINUTES)
+```
+
+Subtracts an interval from a timestamp. Intervals are defined by an integral value and a supported
```
Comment (JimGalasyn) [inline docs/developer-guide/ksqldb-reference/scalar-functions.md:1173]:
```suggestion
Subtracts an interval from a timestamp. Intervals are defined by an integer value and a supported
```

File: docs/reference/sql/time.md [hunk 3]
```diff
@@ -1,7 +1,7 @@
 ### Time units
 
 The following list shows valid time units for the `SIZE`, `ADVANCE BY`,
-`SESSION`, and `WITHIN` clauses.
+`SESSION`, and `WITHIN` clauses, or to pass as intervals in functions.
```
Comment (JimGalasyn) [inline docs/reference/sql/time.md:4]:
```suggestion
`SESSION`, and `WITHIN` clauses, or as intervals in functions.
```

**Review by JimGalasyn — APPROVED** [review]
LGTM, with a few suggestions.

**Comment by blueedgenick** [comment]
Hey @jzaralim - apologies if this was already debated, perhaps in the klip, and I simply missed it - I'm thinking that it might be easier for our users, and more in line with other common systems, to have a `datediff` function rather than specific add/subtract-to-timestamp functions ? In combination with the neat new `interval` building block you have here it can be composed into many interesting permutations. It's pretty common to use this kind of function (and supply a negative interval value to go back in time, positive to go forwards) and of course the naming suggests it may one day work for any kind of date/time data (we can be flexible or not on that as the capabilities get richer).

**Comment by spena** [comment]
@blueedgenick Doesn't `datediff` give a different result than `timestampadd` for instance? The idea of the `timestampadd` is to add units to a timestamp and return a timestamp. But `datediff` returns an integer instead. That seems different, doesn't it?

**Comment by jzaralim** [comment]
@blueedgenick  If you're thinking of a function that generically adds/subtracts an interval of time from a timestamp, then currently there isn't anything stopping an interval from having a negative value, so `timestampadd(date, -1 day)` would actually return the same thing as `timestampsub(date, 1 day)`. I haven't seen any MySQL/Snowflake etc function names that imply this though.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/timestampsub.json [hunk 4]
```diff
@@ -0,0 +1,18 @@
+{
+  "comments": ["tests for timestampsub"],
```
Comment (spena) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/timestampsub.json:2]: Could you add negative tests? Similar to timestampadd()

File: ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4 [hunk 5]
```diff
@@ -297,14 +297,20 @@ primaryExpression
     | MAP '(' (expression ASSIGN expression (',' expression ASSIGN expression)*)? ')'     #mapConstructor
     | STRUCT '(' (identifier ASSIGN expression (',' identifier ASSIGN expression)*)? ')'  #structConstructor
     | identifier '(' ASTERISK ')'                              		                        #functionCall
-    | identifier '(' (expression (',' expression)* (',' lambdaFunction)*)? ')'            #functionCall
+    | identifier'(' expression ',' intervalExpression ')' 						                    #functionCall
+    | identifier '(' (expression (',' expression)*
+                     ((',' lambdaFunction)* | (',' intervalExpression)*))? ')'            #functionCall
```
Comment (spena) [inline ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:302]: Why are there two different functionCall? Doesn't the 2nd one behave similar to `identifier(expr, interval)`?

Btw, I see other DBs uses an INTERVAL keyword, like `dateadd(time, INTERVAL '5 hours')`. Snowflake, DB2 and Mysql uses the INTERVAL keyword. Should we support it too?

File: ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4 [hunk 6]
```diff
@@ -297,14 +297,20 @@ primaryExpression
...
     | value=primaryExpression '[' index=valueExpression ']'                               #subscript
     | identifier                                                                          #columnReference
     | identifier '.' identifier                                                           #qualifiedColumnReference
     | base=primaryExpression STRUCT_FIELD_REF fieldName=identifier                        #dereference
     | '(' expression ')'                                                                  #parenthesizedExpression
     ;
 
+intervalExpression
+    : valueExpression windowUnit
```
Comment (spena) [inline ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:311]: I'm not sure about using `valueExpression` for intervals. We should support only numbers here, right?

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/AstBuilderTest.java [hunk 7]
```diff
@@ -393,7 +395,85 @@ public void shouldNotBuildLambdaFunctionNotLastArguments() {
...
+                )
+            ),
+            Optional.empty())
+    ))));
+  }
+
+  @Test
+  public void shouldBuildMultipleIntervalExpression() {
+    // Given:
+    final SingleStatementContext stmt = givenQuery("SELECT TIMESTAMPADD(Col6, 5 HOURS, 6 MINUTES) FROM TEST1;");
```
Comment (spena) [inline ksqldb-parser/src/test/java/io/confluent/ksql/parser/AstBuilderTest.java:429]: I'm not sure about using multiple intervals this way. I've seen other DBs uses one single expression for multiple intervals, like:
```
timestampadd(col6, '5 hours 6 minutes');
```

Any reason not to support it that way? I don't see other place in ksql that uses it like that.

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/AstBuilderTest.java [hunk 8]
```diff
@@ -393,7 +395,85 @@ public void shouldNotBuildLambdaFunctionNotLastArguments() {
...
+  @Test
+  public void shouldNotBuildIntervalExpressionNotLastArgument() {
+    // Given:
+    final Exception e = assertThrows(
+        ParseFailedException.class,
+        () -> givenQuery("SELECT TIMESTAMPADD(5 HOURS, Col6) FROM TEST1;")
+    );
+
+    // Then:
+    assertThat(e.getMessage(), containsString("no viable alternative at input 'TIMESTAMPADD(5 HOURS"));
```
Comment (spena) [inline ksqldb-parser/src/test/java/io/confluent/ksql/parser/AstBuilderTest.java:464]: Is there a way to change this error message? `no viable alternative at ...` does not look helpful for users. If the problem is generic for other functions, then we can ignore it for now.

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/AstBuilderTest.java [hunk 9]
```diff
@@ -393,7 +395,85 @@ public void shouldNotBuildLambdaFunctionNotLastArguments() {
...
+  @Test
+  public void shouldNotBuildLambdaAndInterval() {
+    // Given:
+    final Exception e = assertThrows(
+        ParseFailedException.class,
+        () -> givenQuery("SELECT TRANSFORM_ARRAY(Col4, X => X + 5, 5 HOURS) FROM TEST1;")
+    );
+
+    // Then:
+    assertThat(e.getMessage(), containsString("mismatched input '5' expecting {"));
```
Comment (spena) [inline ksqldb-parser/src/test/java/io/confluent/ksql/parser/AstBuilderTest.java:476]: Any way to change the error message? Users may get confused about which `5` parameter is.

File: ksqldb-udf/src/main/java/io/confluent/ksql/schema/ksql/types/SqlInterval.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  @Override
+  public String toString(final FormatOptions formatOptions) {
+    return toString();
+  }
+
+  @Override
+  public String toString() {
+    return "INTERVAL";
```
Comment (spena) [inline ksqldb-udf/src/main/java/io/confluent/ksql/schema/ksql/types/SqlInterval.java:37]: Is this `INTERVAL` string printed somewhere? I wonder if you were adding an INTERVAL keyword, but I didn't not see it.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/interval.json [hunk 11]
```diff
@@ -0,0 +1,72 @@
+{
+  "comments": ["tests to ensure that interval is not usable in places it should not be used"],
```
Comment (spena) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/interval.json:2]: How can these intervals use in other places if the `intervalExpression` is not specified anywhere else in SqlBase.g4? I see it is only available for `identifier(...)` which are function calls.

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/expression/tree/IntervalExpression.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  public IntervalExpression(
+      final Optional<NodeLocation> location,
+      final Expression expression,
+      final TimeUnit timeUnit
+  ) {
+    super(location);
+    this.expression = Objects.requireNonNull(expression, "left");
+    this.timeUnit = Objects.requireNonNull(timeUnit, "right");
```
Comment (spena) [inline ksqldb-execution/src/main/java/io/confluent/ksql/execution/expression/tree/IntervalExpression.java:38]:
```suggestion
    this.expression = Objects.requireNonNull(expression, "expression");
    this.timeUnit = Objects.requireNonNull(timeUnit, "timeUnit");
```

**Review by spena — COMMENTED** [review]
Cool, thanks @jzaralim. I left some comments. Btw, I'm not sure if using interval as another type. Is there another way to avoid that? Can we specify the right syntax in SqlBase.g4 instead?

Btw, I looked at other DBs, and they use the `INTERVAL` keyword in their interval parameters. For instance, `date_add(time, INTERVAL '5 hours 5 minutes')`. Snowflake, Postgres and Mysql support it that way. Do you think we should do it too?

**Comment by jzaralim** [comment]
> Cool, thanks @jzaralim. I left some comments. Btw, I'm not sure if using interval as another type. Is there another way to avoid that? Can we specify the right syntax in SqlBase.g4 instead?
> 
> Btw, I looked at other DBs, and they use the `INTERVAL` keyword in their interval parameters. For instance, `date_add(time, INTERVAL '5 hours 5 minutes')`. Snowflake, Postgres and Mysql support it that way. Do you think we should do it too?

The main reason interval is its own type is because we need the UDFs to be able to differentiate between regular integers and intervals. `SqlLambda` exists now though, so it makes sense to refactor intervals to follow the same pattern.

I'm not too sure what purpose the `INTERVAL` keyword would have, but since it is a well known syntax, let's do that too. It should be a simple change anyway.

File: ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4 (hunk 6 above)
Comment (jzaralim) [reply ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:311]: The reason was to support expressions like `4 + 4 DAYS` or `COL0 MINUTES`

**Comment by jzaralim** [comment]
@spena I've changed the grammar to use a separate time unit instead of the interval type from before. I will be rebasing this after #7103 merges though, since it touches a lot of the same code.

File: ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4 [hunk 13]
```diff
@@ -297,14 +297,19 @@ primaryExpression
     | MAP '(' (expression ASSIGN expression (',' expression ASSIGN expression)*)? ')'     #mapConstructor
     | STRUCT '(' (identifier ASSIGN expression (',' identifier ASSIGN expression)*)? ')'  #structConstructor
     | identifier '(' ASTERISK ')'                              		                        #functionCall
-    | identifier '(' (expression (',' expression)* (',' lambdaFunction)*)? ')'            #functionCall
+    | identifier '(' (functionArgument (',' functionArgument)* ((',' lambdaFunction)*))? ')' #functionCall
```
Comment (spena) [inline ksqldb-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4:300]: Are the extra () in the lambdaFunction necessary?

File: ksqldb-parser/src/test/java/io/confluent/ksql/parser/AstBuilderTest.java [hunk 14]
```diff
@@ -396,6 +398,29 @@ public void shouldNotBuildLambdaFunctionNotLastArguments() {
     assertThat(e.getMessage(), containsString("mismatched input '=>' expecting {',', ')'}"));
   }
 
+  @Test
+  public void shouldBuildIntervalUnit() {
+    // Given:
+    final SingleStatementContext stmt = givenQuery("SELECT TIMESTAMP_ADD(MINUTES, 5, Col4) FROM TEST1;");
```
Comment (spena) [inline ksqldb-parser/src/test/java/io/confluent/ksql/parser/AstBuilderTest.java:404]: Isn't it `TIMESTAMPADD`?

**Review by spena — APPROVED** [review]
LGTM. Thanks @jzaralim

**Pull request merged**